steps/test_expval.py

```diff
@@ -0,0 +1,173 @@
+# Copyright 2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This module contains tests locally checking that the functionality of the
+expval step is correct. Running the test cases requires the related packages to
+be installed locally.
+"""
+import math
+import os
+import json
+import numpy as np
+
+import pytest
+from qiskit import IBMQ
+import pennylane as qml
+
+import expval
+
+exact_devices = [
+    '{"module_name": "qeforest.simulator", "function_name": "ForestSimulator", "device_name": "wavefunction-simulator"}',
+    '{"module_name": "qeqiskit.simulator", "function_name": "QiskitSimulator", "device_name": "statevector_simulator"}',
+    '{"module_name": "qequlacs.simulator", "function_name": "QulacsSimulator"}',
+]
+
+sampling_devices = [
+    '{"module_name": "qeforest.simulator", "function_name": "ForestSimulator", "device_name": "wavefunction-simulator", "n_samples": 10000}',
+    '{"module_name": "qeqiskit.simulator", "function_name": "QiskitSimulator", "device_name": "qasm_simulator", "n_samples": 10000}',
+    '{"module_name": "qequlacs.simulator", "function_name": "QulacsSimulator", "n_samples": 10000}',
+]
+
+analytic_tol = 10e-10
+
+# The tolerance for sampling is expected to be higher
+tol = 10e-2
+
+
+@pytest.mark.parametrize("backend_specs", exact_devices)
+class TestExpvalExact:
+    """Tests getting the expecation value of circuits on devices that support
+    exact computations."""
+
+    @pytest.mark.parametrize("op", ['["[Z0]"]', '["[Z1]"]', '["[Z2]"]', '["[]"]'])
+    def test_only_measure_circuit(self, op, backend_specs, monkeypatch):
+        """Tests that the correct result in obtained for a circuit that only
+        contains measurements."""
+        lst = []
+
+        only_measure_qasm = 'OPENQASM 2.0;\ninclude "qelib1.inc";\nqreg q[3];\ncreg c[3];\n'
+
+        monkeypatch.setattr(expval, "save_list", lambda val, name: lst.append(val))
```

Comment:
With Orquestra, at the end of each step we'd serialize the result by dumping it into a `json` file. This is done using the internal `save_list` function in the `run_circuit_and_get_expval` step.

The monkeypatching disables calling this function and instead passes it into a local list so that the value can be checked within the test.

Reply:
Maybe worth adding a comment, super helpful thanks Antal!

Reply:
Sure thing! Added.